Spigot-Server-Patches/0576-Optimize-Villagers.patch

```diff
@@ -39,29 +39,47 @@ So we cache that and only rebuild it if professions change, which should be neve
+@@ -16,7 +16,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+     private final VillagePlaceType b;
+     private final MemoryModuleType<GlobalPos> c;
+     private final boolean d;
+-    private final Optional<Byte> e;
++    private final Optional<Byte> e; private Optional<Byte> getPoiAcquisitionEvent() { return this.e; } // Paper - OBFHELPER
+     private long f;
+     private final Long2ObjectMap<BehaviorFindPosition.a> g;
+ 
+@@ -48,7 +48,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+         if (this.d && entitycreature.isBaby()) {
+             return false;
+         } else if (this.f == 0L) {
+-            this.f = entitycreature.world.getTime() + (long) worldserver.random.nextInt(20);
++            this.f = entitycreature.world.getTime() + (long) java.util.concurrent.ThreadLocalRandom.current().nextInt(20); // Paper
+             return false;
+         } else {
+             return worldserver.getTime() >= this.f;
+@@ -56,12 +56,57 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+     }
  
      protected void a(WorldServer worldserver, EntityCreature entitycreature, long i) {
-         this.f = 0;
--        this.d = worldserver.getTime() + (long) worldserver.getRandom().nextInt(20);
-+        this.d = worldserver.getTime() + (long) java.util.concurrent.ThreadLocalRandom.current().nextInt(20); // Paper
-         VillagePlace villageplace = worldserver.B();
-+
+-        this.f = i + 20L + (long) worldserver.getRandom().nextInt(20);
++        this.f = i + 20L + (long) java.util.concurrent.ThreadLocalRandom.current().nextInt(20); // Paper
+         VillagePlace villageplace = worldserver.y();
+ 
 +        // Paper start - replace implementation completely
-+        BlockPosition blockposition2 = new BlockPosition(entitycreature);
++        BlockPosition blockposition2 = new BlockPosition(entitycreature.locX(), entitycreature.locY(), entitycreature.locZ());
 +        int dist = 48;
 +        int requiredDist = dist * dist;
 +        int cdist = Math.floorDiv(dist, 16);
-+        Predicate<VillagePlaceType> predicate = this.a.c();
++        Predicate<VillagePlaceType> predicate = this.b.getPredicate();
 +        int maxPoiAttempts = 4;
 +        int poiAttempts = 0;
 +        OUT:
 +        for (ChunkCoordIntPair chunkcoordintpair : MCUtil.getSpiralOutChunks(blockposition2, cdist)) {
 +            for (int i1 = 0; i1 < 16; i1++) {
-+                java.util.Optional<VillagePlaceSection> section = villageplace.getSection(SectionPosition.a(chunkcoordintpair, i1).v());
++                java.util.Optional<VillagePlaceSection> section = villageplace.getSection(SectionPosition.a(chunkcoordintpair, i1).asLong());
 +                if (section == null || !section.isPresent()) continue;
 +                for (java.util.Map.Entry<VillagePlaceType, java.util.Set<VillagePlaceRecord>> e : section.get().getRecords().entrySet()) {
```

Comment:
Please use readable names; `e` tells us nothing, especially in the middle of obfuscated code.